### PR TITLE
client: support configuring backoff

### LINKED ISSUE #8047: Introduce the backoff mechanism for gRPC client
## Development Task

- Refactor `tsoClient` to make adopting the later changes easier.
	- [x] #8121
	- [x] #8134
- Introduce the backoff mechanism into the critical paths.
	- [X] #8123
	- [ ] `GetTS` and `GetLocalTS`.
	- [ ] Member update.
- Smooth the TSO internal dispatch to reduce the failure rate.
	- [x] #7997

### PR DESCRIPTION
**Pull request by rleungx**
### What problem does this PR solve?

Issue Number: Close #8047

### What is changed and how does it work?

```commit-message
```

### Check List

Tests

- Unit test

### Release note

```release-note
None.
```

### CODE REVIEW

File: server/grpc_service.go [hunk 1]
```diff
@@ -300,9 +300,7 @@ func (s *GrpcServer) GetMinTS(
 		if done, err := limiter.Allow(fName); err == nil {
 			defer done()
 		} else {
-			return &pdpb.GetMinTSResponse{
-				Header: s.wrapErrorToHeader(pdpb.ErrorType_UNKNOWN, err.Error()),
-			}, nil
+			return nil, status.Error(codes.ResourceExhausted, err.Error())
```
Comment (okJiang) [inline server/grpc_service.go:303]: Maybe we can split these updates to another pr, and merge it firstly. It is simple

File: client/pkg/utils/grpcutil/grpcutil.go [hunk 2]
```diff
@@ -41,6 +42,53 @@ const (
 	FollowerHandleMetadataKey = "pd-allow-follower-handle"
 )
 
+// Add retry related CallOption
+type retryCallOption struct {
+	grpc.EmptyCallOption
```
Comment (nolouch) [inline client/pkg/utils/grpcutil/grpcutil.go:47]: why need this?

File: tests/integrations/client/client_test.go [hunk 3]
```diff
@@ -1993,3 +1995,54 @@ func (suite *clientTestSuite) TestBatchScanRegions() {
...
+	bo.SetRetryableChecker(needRetry, true)
+
+	// Initialize the client with context and backoff
+	client, err := pd.NewClientWithContext(ctx, caller.TestComponent, endpoints, pd.SecurityOption{})
+	re.NoError(err)
+
+	// Record the start time
+	start := time.Now()
+
+	ctx = retry.WithBackoffer(ctx, bo)
```
Comment (nolouch) [inline tests/integrations/client/client_test.go:2026]: Does the user need to specify it?

Comment (rleungx) [reply tests/integrations/client/client_test.go:2026]: Yes, but we also can provide a default one if needed.

File: client/pkg/utils/grpcutil/grpcutil.go (hunk 2 above)
Comment (rleungx) [reply client/pkg/utils/grpcutil/grpcutil.go:47]: Otherwise, we need to implement before and after interfaces.

File: client/pkg/retry/backoff.go [hunk 4]
```diff
@@ -191,3 +191,25 @@ func getFunctionName(f any) string {
 	strs := strings.Split(runtime.FuncForPC(reflect.ValueOf(f).Pointer()).Name(), ".")
 	return strings.Split(strs[len(strs)-1], "-")[0]
 }
+
+// Define context key type
+type boCtxKey struct{}
+
+// Key used to store backoffer
+var backofferKey = boCtxKey{}
```
Comment (okJiang) [inline client/pkg/retry/backoff.go:199]: Why using `struct{}`, not string?

File: tests/integrations/client/client_test.go [hunk 5]
```diff
@@ -1993,3 +1995,54 @@ func (suite *clientTestSuite) TestBatchScanRegions() {
...
+	// Call GetRegion and expect it to handle backoff internally
+	_, err = client.GetRegion(ctx, []byte("key"))
+	re.Error(err)
+	// Calculate the elapsed time
+	elapsed := time.Since(start)
+	// Verify that some backoff occurred by checking if the elapsed time is greater than the base backoff
+	re.Greater(elapsed, base, "Expected some backoff to have occurred")
```
Comment (okJiang) [inline tests/integrations/client/client_test.go:2033]: Is it better to compare `total`?

File: client/pkg/retry/backoff.go (hunk 4 above)
Comment (rleungx) [reply client/pkg/retry/backoff.go:199]: I prefer to use struct{}

**Review by okJiang — APPROVED** [review]
